File: day_03/day03.py
#!/usr/bin/python3
# Advent of Code 2019
# Day 3

from typing import Tuple, List

split_data = lambda x: [element.split(",") for element in x.splitlines()]

closest_manhattan_distance = lambda first, second: min(
    (abs(intersection[0]) + abs(intersection[1])
     for intersection in set(first) & set(second) if intersection != (0, 0)))

step_count = lambda first, second: min(
    (first.index(intersection) + second.index(intersection)
     for intersection in set(first) & set(second) if intersection != (0, 0)))


def path(moves: List[str]):
    directions = {"R": 1, "L": -1, "U": 1j, "D": -1j}
    wiring: List[Tuple[int, int]] = [(0, 0)]

    for move in moves:
        [wiring.append((wiring[-1][0] + int(directions[move[0]].real),
                        wiring[-1][1] + int(directions[move[0]].imag)))
         for _ in range(int(move[1:]))]

    return wiring


def unpack_paths(input_data):
    wires = [path(wire) for wire in split_data(input_data)]
    return wires[0], wires[1]


if __name__ == "__main__":
    wires = unpack_paths(open("input").read())
    print(f"Manhattan distance: {closest_manhattan_distance(wires[0], wires[1])}")
    print(f"Step count: {step_count(wires[0], wires[1])}")

File: day_03/day03.pl
#!/usr/bin/perl
# Advent of Code 2019
# Day 3

use strict;
use warnings;
use List::Util qw(min);
use List::MoreUtils qw(first_index);
#use Set::Light;
#use Set::Intersection;
use Array::Utils qw(:all);
use Math::Complex;

sub split_data {
    return map {split /,/} @_;
}

sub closest_manhattan_distance {
    my @first = $_[0];
    my @second = $_[1];
    my @zeros = (0, 0);
    my @intersection = intersect(@first, @second);
    my @minus = array_minus(@intersection, @zeros);
    return min(map {abs($_[0]) + abs($_[1])} @minus);
}

sub step_count {
    my @first = $_[0];
    my @second = $_[1];
    my @zeros = (0, 0);
    my @intersection = intersect(@first, @second);
    my @minus = array_minus(@intersection, @zeros);
    return min(map {my $one = $_[0]; my $two = $_[1];
        firstindex {$_[0] == $one} @first + firstindex{$_[1] == $two} @second} @minus);
}

sub path {
    my %directions = (
        R => 1,
        L => -1,
        U => 1*i,
        D => -1*i
    );
    my @wiring = ([0, 0]);
    
    for my $move (@_) {
        my $vector = substr $move, 0, 1;
        my $magnitude = (substr $move, 1);
        for my $i (0 .. $magnitude) {
            push(@wiring, ($wiring[-1][0] + Re($directions{$vector}), $wiring[-1][1] + Im($directions{$vector})));
        }
    }
    return @wiring;
} 

sub unpack_paths {
    open(my $in, "<", "input") or die "Can't open: $!";
    chomp(my @wires = <$in>);
    close $in;
    return map {split /,/, $_} @wires;
}

my @wires = unpack_paths();
print scalar(@wires), "\n";
print "Manhattan distance: ", closest_manhattan_distance(@wires), "\n";
print "Step count: ", step_count(@wires), "\n";

